EH support in the interpreter

cshung opened the pull request:
This change implements the exception handling support in the interpreter.

This is still a work in progress, right now all test cases are passing locally on Windows x64 Debug.

cshung committed:
EH support in the interpreter

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -4339,8 +4339,34 @@ BOOL InterpreterJitManager::JitCodeToMethodInfo(
+        // before its handler funclet. So the filter end offset is equal to the start offset of the handler funclet.
+        if (IsFilterHandler(&ehClause) && (ehClause.FilterOffset <= relOffset) && (relOffset < ehClause.HandlerStartPC))
+        {
+            // Filter handlers are not funclets, but we need to return the start address of the filter code.
```
What are the filter handlers if they are no funclets?

cshung replied:
Oh, that comment was just a typo ... I'll delete the whole line.

jkotas commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -1903,11 +2177,36 @@ int InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
+    // and the blocks of others.
+    //
+    // The reason of doing that is that the runtime requires the funclets to be separated from
+    // the body so that funclets are never inside try
```
Funclets can be inside a `try`. For example:
```
try // TryA
{
    try // TryB
    {
        ...
    }
    finally
    {
        // This is a funclet that is inside (or more precisely - covered by) TryA
    }
}
catch
{
    ...
}
```

jkotas replied:
The regular JIT duplicates the EH clauses as necessary to move funclets out-of-file while keeping them covered by try clauses. My example has two EH clauses in IL, but it has 4 EH clauses for the generated code due to this duplication.

Is the interpreter going to do the same?

janvorli replied:
> The regular JIT duplicates the EH clauses as necessary to move funclets out-of-file while keeping them covered by try clauses.

It does not anymore. @filipnavara has removed creating them recently as the new EH didn't use them.

janvorli replied:
When exception escapes a funclet, we have the "collided handling" mechanism that restores the stack frame iterator from the previous ExInfo and continues looking for handler where the previous EH left it. This is the same way as we have always had in nativeAOT.

jkotas replied:
Ah ok, I have not realized that we got rid of the duplication completely. I thought that there is still some left.

cshung committed:
Fix clang errors

cshung commented:
The test failure under Windows/x64/Checked is somewhat understood. It looks like the C# compiler is optimizing away a `LEAVE` opcode that this change is depending on to invoke a finally. I will think about what to do about that.

janvorli commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1213,6 +1370,7 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
     }
     catch (const ResumeAfterCatchException& ex)
     {
+        GCX_COOP_NO_DTOR();
```
We should instead fix the source of the problem - remove the switch to preemptive mode from the `CallDescrWorkerUnwindFrameChainHandler`. I believe we don't need to have it there.

janvorli commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1200,11 +1285,83 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
+                    if (non_exceptional_finally_count > 0)
+                    {
+                        // This opcode returns from the finally block
+                        non_exceptional_finally_count -= 1;
```
Will this count be properly maintained in case one of the non-exceptionally called finallys throws and the exception is caught in the same parent frame?

janvorli committed:
Implement GetFuncletStartOffsets for interpreter

An unrelated failure in the CI has shown that we are also missing
InterpreterCodeManager::GetFuncletStartOffsets that is used when
reporting an unhandled exception to the debugger / Watson.
This change adds that.

cshung commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -64,6 +107,11 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
             // keep it for such purposes until we don't need it anymore.
             pFrame->ip = (int32_t*)ip;
 
+#ifdef DEBUG
```
I kind of like this trace, it helped me to debug countless time. It makes understanding what happened in the past so much easier.

janvorli replied:
Ok. Let's keep it in for now.

cshung committed:
Disable failing test cases caused by Roslyn optimization

BrzVlad commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -3257,27 +3606,107 @@ int InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
+                    InterpInst* next = AddIns(INTOP_NOP);
+                    int ilOffset = (int)(m_ip - m_pILCode);
+                    int target = (opcode == CEE_LEAVE) ? ilOffset + 5 + *(int32_t*)(m_ip + 1) : (ilOffset + 2 + (int8_t)m_ip[1]);
+                    if (m_pCBB->finally_index != -1)
```
A leave opcode can jump over multiple protected blocks. It seems to me that here you only have support for invoking a single finally clause, when it might be the case that you need to invoke more ?

cshung replied:
This is something I am working on right now. We need to handle "interesting" usage of the leave opcode that exit more than one protected region. This never happens with the C# compiler that generate debug IL and so I wasn't aware of that earlier.

BrzVlad replied:
This is actually easy to reproduce, for example:
```
Method ()
{
   try {
      try {
         return;
      } finally {
      }
   } finally {
   }
}
```

The return will leave over multiple finally blocks

BrzVlad commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1200,11 +1285,83 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
+                        pFrame->ip = ip + 3;
+                        const int32_t* targetIp = ip + ip[2];
+
+                        // Allocate child frame.
```
I don't understand why do we allocate a new frame here ? It seems like this code is making the opcode implementation more complicated than it needs to be. On mono, each finally has a dedicated offset where it can receive the return address (If this address is 0 then it will return to EH). So this opcode could just set the return ip to some var and then continue with opcode dispatch. Also `non_exceptional_finally_count` would no longer be necessary.

janvorli replied:
To be able to handle throw from finally using the same way as for JIT/AOTed code, even non-exceptionally called finally need to be invoked via a call so that it has its own frame on the stack. The reason is that all funclets are moved out of any try ranges to make throw from funclets behave correctly with the EH algorithm. JIT/AOT does the same thing on coreclr.

BrzVlad replied:
I see. Seems unfortunate that our cost for invoking finally will be significantly higher than for JIT, even in non-eh case. Maybe at some point in the future we could make some changes on the VM side of EH so this is no longer necessary. A finally funclet could make use only of a var containing the return ip, used when reaching endfinally. The EH mechanism could, in theory, obtain this ip and then further determine whether the code is protected, handling it accordingly.

BrzVlad commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1200,11 +1285,83 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
                     break;
                 }
+
+                case INTOP_ENTER_FUNCLET:
```
I don't think there is any reason for `INTOP_ENTER_FUNCLET` to exist and to pass throwable as an argument to interp exec method. When invoking a clause, EH could write directly to the stack offset where the exception object is expected.

cshung replied:
This opcode helped me with a few things:

1. This allow all funclet blocks to have an empty IL evaluation stack to start with - that also made sure all the funclet entry blocks are generated on the first pass without retry. 
2. This allow me to decouple the IR generator from the interpreter by not having to describe where the exception is stored.
3. This allow me to use a local slot to store the exception object instead of a global one, that space is likely reused for the non-exceptional path.

Not sure what is the advantage of doing otherwise.

BrzVlad replied:
In general we should aim to have as few opcodes generated as possible for the sake of performance. In this particular example, given this is only used in the exceptional path for filter and catch funclets, I'd say it is ok to leave as is, given 2 and 3 represent good enough motivation.

BrzVlad commented on src/coreclr/vm/interpexec.h:
```diff
@@ -61,6 +61,18 @@ struct InterpThreadContext
+// In the case of catch, it will return the bytecode address where the execution should resume, or
+// In the case of filter, it will return the decision to either execute on the current handler or continue searching for another handler.
+//
+DWORD_PTR ExecuteInterpretedCode(TransitionBlock* pTransitionBlock, TADDR byteCodeAddr, OBJECTREF throwable, void* pHandler, InterpMethodContextFrame* handlerFrame, bool isFilter);
```
I don't think we should be having so many arguments for `ExecuteInterpretedCode`. For the invoke from EH path it seems cleaner and more perf efficient to have a single struct argument that contains all relevant information necessary for this uncommon path. Then, the typical method call would pass a simple null in this place.

BrzVlad commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -825,16 +849,29 @@ void InterpCompiler::EmitCode()
+    bool insideFunclet = false;
     for (InterpBasicBlock *bb = m_pEntryBB; bb != NULL; bb = bb->pNextBB)
     {
+        bool enteringFunclet = false;
```
Seems weird that there is quite a bit of logic around computing `enteringFunclet` and passing it over, and it is only used for one assertion.

BrzVlad commented on src/coreclr/vm/eetwain.cpp:
```diff
@@ -2144,9 +2144,8 @@ void EECodeManager::UpdateSSP(PREGDISPLAY pRD)
-    // Interpreter-TODO: implement calling the funclet in the intepreted code
-    _ASSERTE(FALSE);
-    return 0;
+    InterpMethodContextFrame* faultingFrame = (InterpMethodContextFrame*)pRD->SP;
```
I thought the plan was be to create a new frame only for filter funclets, because the stack wasn't yet unwound when they are called. For finally and catch funclets I thought we would be able to call them from the same exact frame without making use of this frame var.

janvorli replied:
We have found it would not work with the coreclr. It would cause frames reported by the stack walk to not to be ordered by their stack location anymore. It would result in reporting the frame in whose context the finally was running, skipping several explicit frames like the InterpreterFrame and the pinvoke frame of the CallFinallyFunclet / CallCatchFunclet. Besides the ordering issue, skipping the CallFinallyFunclet / CallCatchFunclet would prevent the EH detecting that the exception has escaped from a funclet. The throw from finally was a case where we have found this was a problem.

janvorli replied:
I am not 100% sure I understand what you meant. The first sequence seemed to indicate that we don't need a separate InterpMethodContextFrame for the funclets, but reading the rest seems the meaning was actually the opposite. If the point was that we can have a separate InterpMethodContextFrame for the funclets, but use the interpreter stack pointer of the parent frame of the funclet, then I agree. I've already mentioned that to Andrew before.

BrzVlad replied:
I initially thought we might share also the `InterpMethodContextFrame`, but my original observation was more about sharing the method \'stack frame\' from the interpreter stack and avoiding the use of a separate base pointer.

>  If the point was that we can have a separate InterpMethodContextFrame for the funclets, but use the interpreter stack pointer of the parent frame of the funclet, then I agree. I've already mentioned that to Andrew before.

Sounds all good then

BrzVlad commented on src/coreclr/interpreter/intops.def:
```diff
@@ -31,6 +31,22 @@ OPDEF(INTOP_MOV_4, "mov.4", 3, 1, 1, InterpOpNoArgs)
 OPDEF(INTOP_MOV_8, "mov.8", 3, 1, 1, InterpOpNoArgs)
 OPDEF(INTOP_MOV_VT, "mov.vt", 4, 1, 1, InterpOpInt)
 
+OPDEF(INTOP_LOAD_I4_I1, "load.i4.i1", 3, 1, 1, InterpOpNoArgs)
```
As with previous discussions on this topic, sounds to me that this frame var indirection is only needed for filter funclets. In this case it doesn't make sense to have so many opcodes, we could replace all of them via a single `load.framevar` and already existing indirect loads.

cshung replied:
I still recall that discussion, although I am not sure if reducing the number of opcodes should be a goal. I think we might likely increase the time spent on executing more opcodes and also making the code a bit more complicated.

I will give that a try to confirm if that change worth it.
